include/class.queue.php

```diff
@@ -1260,6 +1292,53 @@ function save($refetch=false) {
             && $this->sorts->saveAll();
     }
 
+    /**
+     * Fetch a tree-organized listing of the queues. Each queue is listed in
+     * the tree exactly once, and every visible queue is represented. The
+     * returned structure is an array where the items are two-item arrays
+     * where the first item is a CustomQueue object an the second is a list
+     * of the children using the same pattern (two-item arrays of a CustomQueue
+     * and its children). Visually:
+     *
+     * [ [ $queue, [ [ $child, [] ], [ $child, [] ] ], [ $queue, ... ] ]
+     *
+     * Parameters:
+     * $staff - <Staff> staff object which should be used to determine
+     *      visible queues.
+     * $pid - <int> parent_id of root queue. Default is zero (top-level)
+     */
+    static function getHierarchicalQueues(Staff $staff, $pid=0) {
+        $criteria = Q::any(array(
+            'flags__hasbit' => self::FLAG_PUBLIC,
+            'flags__hasbit' => static::FLAG_QUEUE,
+            'staff_id' => $staff->getId(),
+        ));
+
+        if (($teams = $staff->getTeams())) {
+            $criteria->add(Q::all(array(
+                'flags__hasbit' => self::FLAG_VISIBLE_TEAM,
+                'staff_id__in' => $teams,
```

Comment:
I have a feeling we need `ost_queue_access` table to truly handle sharing. Think ability to share / restrict queue access to teams, depts and groups as part of sharing.